include/dali/core/exec/tasking/task.h

```diff
@@ -0,0 +1,522 @@
+        auto &&results = f(std::forward<Args>(args)...);
+        UnpackResults<0>(results_, std::move(results));
+      } else {
+        assert(!"Internal error - the output type should have been rejected earlier.");
+      }
+    } catch (...) {
+      auto ex = std::current_exception();
+      for (auto &r : results_)
+        r->SetException(ex);
+    }
+  }
+
+  void InitResults(int n) {
+    assert(n > 0);
+    results_.resize(n);
+    for (auto &r : results_)
+      r = std::make_shared<TaskResult>();
+  }
+
+  template <typename R>
+  void CheckResultType(int num_results) {
+    using Result = std::remove_reference_t<R>;
+    if constexpr (detail::is_iterable_v<Result>) {
+      return;
+    } else if constexpr (detail::is_tuple_v<Result>) {
+      if (std::tuple_size_v<Result> != num_results)
+        throw std::logic_error("The output tuple has a different size than "
+                               "the declared number of task's ouputs.");
+    } else {
+      throw std::invalid_argument("The result of the function is neither iterable nor a tuple "
+                                  "and cannot be used to obtain multiple output values.");
+    }
+  }
+
+
+  template <typename F>
+  void CheckFuncResultType(int num_results, F &&function) {
+    using Func = std::remove_reference_t<F>;
+    if constexpr (std::is_invocable_v<Func, Task *>)
+      CheckResultType<decltype(function(this))>(num_results);
+    else
+      CheckResultType<decltype(function())>(num_results);
+  }
+
+ public:
+  template <typename F>
+  Task(int num_results, F &&function, double priority = 0) {
+    using Func = std::remove_reference_t<F>;
+    priority_ = priority;
+    results_.Init(num_results);
+
+    if (num_results == ScalarResult) {
+      wrapped_ = [f = std::forward<F>(function)](Task *t) {
+        using Func = std::remove_reference_t<F>;
+        if constexpr (std::is_invocable_v<Func, Task *>)
+          t->SetResult(f, t);
+        else if constexpr (std::is_invocable_v<Func>)
+          t->SetResult(f);
+      };
+    } else {
+      CheckFuncResultType(num_results, std::forward<F>(function));
+      wrapped_ = [f = std::forward<F>(function)](Task *t) {
+        if constexpr (std::is_invocable_v<Func, Task *>)
+          t->SetResults(f, t);
+        else if constexpr (std::is_invocable_v<Func>)
+          t->SetResults(f);
+      };
+    }
+  }
+
+  template <typename F>
+  explicit Task(F &&function, double priority = 0)
+  : Task(ScalarResult, std::forward<F>(function), priority) {}
+
+  /** Creates a task with a scalar result
+   *
+   * @param function    the callable object that defines the task; it can return any type
+   * @param priority    the priority with which the task will be popped by the scheduler, once ready
+   */
+  template <typename F>
+  static SharedTask Create(F &&function, double priority = 0) {
+    return std::make_shared<Task>(std::forward<F>(function), priority);
```

Comment:
```suggestion
    return Create(ScalarResult, F, priority);
``?

Reply:
Then it would bypass the matching constructor. I can remove it if you think it's better.

Reply:
Up to you.